[sw, tock] Generate DIF bindings and link against the OT libraries

silvestrst opened the pull request:
The downstream Tock OS drivers are going to use OT DIF, libbase, and possibly some other libraries. In order to do that, it needs to link against these libraries. This is achieved through producing bindings for the symbols in these libraries and linking against them. For more information see the commit messages.

Currently DIF bindings are in a single `.rs` file, generated from all required libraries. We can improve that in follow up changes, but it is sufficient for now.

[It seems that regular expressions are supported](https://docs.rs/bindgen/0.33.0/bindgen/struct.Builder.html), so we can do clever things with bindgen to generate friendly bindings.

This change depends on a more fundamental #2119 change.

**Please only review the following changes, and beforehand have a look at the corresponding issue (#2139 ) - which has complete outline of the problem:
[sw, tock] Generate Rust bindings to DIF headers
[sw, tock] Link Tock 'dif' crate against OT DIF libraries**

mcy commented on sw/device/tock/boards/opentitan/src/counter_demo.rs:
```diff
@@ -4,6 +4,8 @@
 
 //! A simple alarm implementation to display a binary counter over GPIO.
 
+use tock::kernel as kernel;
```
You can just write `use tock::kernel;`. This applies everywhere else.

silvestrst replied:
Hmmm, I am sure I had issues where in the functions `kernel::<something> `was used, but I will double check - if there are no issues I will change it up.

silvestrst replied:
Thank you @mcy , fixing it up now!

silvestrst replied:
Done.

mcy commented on sw/device/tock/boards/opentitan/src/main.rs:
```diff
@@ -24,7 +29,7 @@ pub mod io;
     [None, None, None, None];
 
-static mut CHIP:Option<&'static ibex::chip::Ibex> = None;
+static mut CHIP:Option<&'static earlgrey::chip::Earlgrey> = None;
```
Could you run rustfmt over your change?

mcy commented on sw/device/tock/boards/opentitan/src/main.rs:
```diff
@@ -140,7 +145,7 @@ pub unsafe fn reset_handler() {
     DynamicDeferredCall::set_global_instance(dynamic_deferred_caller);
 
-    let chip = static_init!(ibex::chip::Ibex, ibex::chip::Ibex::new());
+    let chip = static_init!(earlgrey::chip::Earlgrey, earlgrey::chip::Earlgrey::new());
```
Consider `use earlgrey::chip::Earlgrey;`?

mcy commented on sw/device/tock/chips/earlgrey/src/chip.rs:
```diff
@@ -0,0 +1,269 @@
+
+    unsafe fn handle_plic_interrupts() {
+        while let Some(interrupt) = plic::next_pending() {
+            match interrupt {
+                interrupts::UART_TX_WATERMARK..=interrupts::UART_RX_PARITY_ERR => {
+                    uart::UART0.handle_interrupt()
+                }
+                int_pin @ interrupts::GPIO_PIN0..=interrupts::GPIO_PIN31 => {
+                    let pin = &gpio::PORT[(int_pin - interrupts::GPIO_PIN0) as usize];
+                    pin.handle_interrupt();
+                }
+                _ => debug!("Pidx {}", interrupt),
+            }
+            plic::complete(interrupt);
+        }
```
This code could be made more readable if you add `use interrupts::*;` right before the match expression to bring all of the constants into scope. As it stands, the patterns are a bit to long and the `..=` is hard to see.

silvestrst replied:
This is directly copied code from upstream, the plan is to clean everything up in follow-up commits (probably when we get to refactor plic).

However, this tip is very useful, I didn't realise that you can simply do `*` to pull in all the components in scope. Thank you!

mcy commented on sw/device/tock/chips/earlgrey/src/chip.rs:
```diff
@@ -0,0 +1,269 @@
+                interrupts::UART_TX_WATERMARK..=interrupts::UART_RX_PARITY_ERR => {
+                    uart::UART0.handle_interrupt()
+                }
+                int_pin @ interrupts::GPIO_PIN0..=interrupts::GPIO_PIN31 => {
```
If `interrupt` is `Copy`, consider just using that instead of creating a binding here. `var @ pat` is only really useful for non-`Copy` types.

mcy commented on sw/device/tock/chips/earlgrey/src/chip.rs:
```diff
@@ -0,0 +1,269 @@
+    pub unsafe fn new() -> Earlgrey {
+        Earlgrey {
+            userspace_kernel_boundary: SysCall::new(),
+            pmp: rv32i::pmp::PMPConfig::new(4),
```
Could you pull `4` into a constant?

mcy commented on sw/device/tock/chips/earlgrey/src/chip.rs:
```diff
@@ -0,0 +1,269 @@
+    match exception {
+        mcause::Exception::UserEnvCall | mcause::Exception::SupervisorEnvCall => (),
+
+        mcause::Exception::InstructionMisaligned
+        | mcause::Exception::InstructionFault
+        | mcause::Exception::IllegalInstruction
+        | mcause::Exception::Breakpoint
+        | mcause::Exception::LoadMisaligned
+        | mcause::Exception::LoadFault
+        | mcause::Exception::StoreMisaligned
+        | mcause::Exception::StoreFault
+        | mcause::Exception::MachineEnvCall
+        | mcause::Exception::InstructionPageFault
+        | mcause::Exception::LoadPageFault
+        | mcause::Exception::StorePageFault
+        | mcause::Exception::Unknown => {
```
Consider using just `_ => ` instead.

mcy commented on sw/device/tock/chips/earlgrey/src/chip.rs:
```diff
@@ -0,0 +1,269 @@
+            handle_interrupt(interrupt);
+        }
+        _ => {
+            panic!("unexpected non-interrupt\n");
```
What's the `\n'` for?

mcy commented on sw/device/tock/chips/earlgrey/src/lib.rs:
```diff
@@ -0,0 +1,16 @@
+//! Drivers and chip support for the Ibex soft core.
+
+#![feature(asm, concat_idents, const_fn, naked_functions)]
+#![feature(in_band_lifetimes)]
```
Consider not using this feature, since it's not mainline Rust and I'm working to remove it from Tock.

mcy commented on sw/device/tock/chips/earlgrey/src/lib.rs:
```diff
@@ -0,0 +1,16 @@
+#![feature(asm, concat_idents, const_fn, naked_functions)]
+#![feature(in_band_lifetimes)]
+#![no_std]
+#![crate_name = "earlgrey"]
+#![crate_type = "rlib"]
```
Is there a reason you need these? Normally Cargo provides these to the compiler.

mcy commented on sw/device/tock/chips/earlgrey/src/plic.rs:
```diff
@@ -0,0 +1,95 @@
+
+/// Clear all pending interrupts.
+pub unsafe fn clear_all_pending() {
+    let _plic: &PlicRegisters = &*PLIC_BASE;
```
It's not clear what this is doing. Why are you creating a binding and not using it? Is there code missing. or dos the deref operation have a side-effect?

mcy commented on sw/device/tock/chips/earlgrey/src/plic.rs:
```diff
@@ -0,0 +1,95 @@
+pub unsafe fn has_pending() -> bool {
+    let plic: &PlicRegisters = &*PLIC_BASE;
+
+    plic.pending.iter().fold(0, |i, pending| pending.get() | i) != 0
```
I think that it's best to avoid iterator combinators for something like this, which can be much more simply written as
```rust
for pending in plic.pending.iter() {
  if pending.get() != 0 {
    return true
  }
}

false
```

mcy commented on sw/device/tock/chips/opentitan_common/src/gpio.rs:
```diff
@@ -0,0 +1,253 @@
+        upper: &ReadWrite<u32, mask_half::Register>,
+    ) {
+        let shift = field.shift;
+        let bit = if val { 1u32 } else { 0u32 };
```
You can just cast `bool` to `u32`; the language promises this conversion is `1` for `true` and `0` for `false`.

mcy replied:
Unlike in `C`, you don't need to put `u32` everywhere to make numbers unsigned. The language will pick the appropriate type based on usage, which works out for numbers.

mcy commented on sw/device/tock/chips/opentitan_common/src/gpio.rs:
```diff
@@ -0,0 +1,253 @@
+        if regs.intr_state.is_set(pin) {
+            regs.intr_state.modify(pin.val(1));
+            self.client.map(|client| {
+                client.fired();
```
You can just write `|client| client.fired()`.

mcy commented on sw/device/tock/chips/opentitan_common/src/lib.rs:
```diff
@@ -0,0 +1,11 @@
+
+#![feature(asm, concat_idents, const_fn, core_intrinsics)]
+#![feature(in_band_lifetimes)]
+#![feature(exclusive_range_pattern)]
```
Consider avoiding `features` that you don't truly need, especially since I'm working to remove all of them from Tock.

mcy commented on sw/device/tock/chips/opentitan_common/src/uart.rs:
```diff
@@ -0,0 +1,320 @@
+
+    pub fn transmit_sync(&self, bytes: &[u8]) {
+        let regs = self.registers;
+        for b in bytes.iter() {
```
You can get an iterator of type `u8` instead of `&u8` by adding `.copied()`, which copies each iterated element.

mcy commented on sw/device/tock/chips/opentitan_common/src/uart.rs:
```diff
@@ -0,0 +1,320 @@
+    }
+}
+
+/* UART receive is not implemented yet, mostly due to a lack of tests avaliable */
```
Avoid `/*` comments; use `//` comments instead.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+// Copyright lowRISC contributors.
```
Please add a comment explaining how this file works and what it expects as input. It's very clearly a file that assumes Meson is running.

silvestrst replied:
Apologies, missed this comment - in process of adding it.

silvestrst replied:
Done, also ran rustfmt.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+    }
+}
+
+fn generate_bindings(source_root: &String, build_root: &String, out_dir: &String) {
```
Since you use these exclusively as paths, consider accepting `impl AsRef<Path>` instead; you can then do `source_root.as_ref().join("blah");` below.

silvestrst replied:
It then complains about `.clang_args(&["-I", &source_root])`, so I kept arguments as `&str`.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+    libraries_link(mmio_lib_path.to_str().unwrap(), base_libs);
+}
+
+fn libraries_link(path: &str, libs: &[&str]) -> () {
```
`-> ()` is redundant. You can drop it.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+    generate_bindings(&source_root, &build_root, &out_dir);
+
+    let dif_lib_path = Path::new(&build_root).join("sw/device/lib/dif");
+    let dif_libs = &["uart_ot"];
```
I would consider inlining this variable into the call below. Creating a variable for a single use like this isn't very common in Rust.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+use std::path::Path;
+
+fn main() {
+    let source_root = env::var("MESON_SOURCE_ROOT").unwrap();
```
Avoid `unwrap`, since the error it gives is mostly useless, and use `expect` instead.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+
+extern crate bindgen;
+
+use bindgen::builder;
```
I would avoid a line like this. `builder()` below looks pretty confusing, since it's not clear where it came from, and I think I would have written `bindgen::builder()`. In general, it's best to only import stuff you wind up using a lot.

alphan replied:
While I understand that this is a small file and this line doesn't cause a lot of ambiguity, I agree with @mcy. From https://doc.rust-lang.org/book/ch07-04-bringing-paths-into-scope-with-the-use-keyword.html:
> Bringing the function’s parent module into scope with use so we have to specify the parent module when calling the function makes it clear that the function isn’t locally defined while still minimizing repetition of the full path.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+
+use bindgen::builder;
+use std::env;
+use std::fs::{self};
```
`::{self}` is redundant.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+    for file in files {
+        let path = file.unwrap().path();
+        if let Some(e) = path.extension() {
+            let path = path.to_str().unwrap();
```
Maybe don't name this the same as `path` above?

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,62 @@
+    let files = fs::read_dir(dif_path).unwrap();
+    for file in files {
+        let path = file.unwrap().path();
+        if let Some(e) = path.extension() {
```
Consider `e` -> `extension`.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/src/lib.rs:
```diff
@@ -0,0 +1,14 @@
+#![crate_type = "rlib"]
+
+#[allow(nonstandard_style)]
+pub mod dif_bind;
```
Consider instead:
```
pub mod dif_bind {
  include!(...);
}
```

alphan replied:
+1 to this and the comment above.

silvestrst replied:
Sounds good, thanks.

silvestrst replied:
Honestly I find it really confusing when in the [documentation](https://doc.rust-lang.org/std/macro.include.html) for `std`:
`Macro std::include`
you click on the [src](https://doc.rust-lang.org/src/core/macros/mod.rs.html#1223-1226), and it actually turns out to be `core`.

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/src/lib.rs:
```diff
@@ -0,0 +1,14 @@
+
+#[allow(nonstandard_style)]
+pub mod dif_bind;
+#[allow(nonstandard_style)]
```
I would move this into the module file, and then mark each type explicitly with it.

silvestrst replied:
What is your reasoning here? In case the module file contains some standard types?

silvestrst replied:
Done.

mcy commented on sw/device/tock/dif/src/lib.rs:
```diff
@@ -0,0 +1,14 @@
+// Copyright lowRISC contributors.
```
Is this library *just* generated binding symbols? If so, I would consider renaming it `dif_bindings` to make it clear that it does not export a nice Rust API.

silvestrst replied:
@mcy thank you for reviewing this, I will try to address these comments tomorrow.

> Throughout: please document every function and variable, as we would in C. Tock has a bad habit of not doing this, but I don't think that's reason to not write the comments.

90% of the code is not written by me, I think the cleanup can be done later in follow-up PRs. At this point most of the code has been taken from the upstream, and is expected to gradually divert.

lenary commented:
@mcy Thank you for the thorough review. 

This PR builds upon #2119 (a dependency that github makes it impossible to show). That PR is an almost direct vendoring of lots of the existing OT-related tock sources (modulo having to change some module paths, for which your suggestions are helpful). 

This PR then builds on that to start running bindgen over the DIFs so we can make Tock wrappers for them. Thanks for the comments on build.rs, which I think is the core part of this PR.

These two PRs, together, should leave us ready for updating and improving tock as we slowly replace the existing HIL implementations with DIF-based implementations.

I hope this clarifies what this PR hopes to achieve, and our approach as we spin-up downstream Tock development. I'd be happy to talk the approach over if that would be helpful.

mcy commented:
@lenary: Yikes. I had an inkling that this code looked awfully like something lifted out of Tock, but I assumed that Silvestrs had just based off of what they did. It's alright though, we don't need to address the problems in the upstream code right now.

I think that all sounds good. I don't completely understand the exact build mechanism, though. I don't need an explanation right now, but I think it should be explained in the "how to build OT" doc I shared around a while ago at some point.

lenary commented:
@mcy The build mechanism today is 'there's a custom action in meson, it's not enabled by default'. 

I don't think we don't yet have a clear idea how we go about switching this on by default yet, as we need to work out how to enable Rust in CI (though enabling it for the non-software devs is probably a more complex question). 

I'm keen we get these first two PRs in and settle those issues once we have something more specific to poke at. I don't think anything in these first two PRs is going to paint us into a corner on those questions about the build system and how we update to upstream, but if you do have concerns, do let us know.

silvestrst commented:
> It's alright though, we don't need to address the problems in the upstream code right now.

Also a thing to remember is that all this code is not upstream `HEAD`, I think I should open a tracking issue and list everything that has to be done.

@mcy your comments I will use in my UART refactoring. So that is the idea - we are taking everything as it is, and then as we refactor every peripheral driver to use DIFs, we also clean-up the legacy code for that driver. It is easier and more practical than to do a massive refactoring in one go.

Again, thank you for your review. I think we should use the comments you made as a reference for all the follow up cleanup we do.

alphan commented on sw/device/tock/dif/build.rs:
```diff
@@ -0,0 +1,47 @@
+        }
+    }
+
+    let out_file = Path::new(&out_dir).join("ot_dif_bindings.rs");
+    binding_builder.generate().unwrap().write_to_file(out_file).unwrap();
```
Cool, I'm curious what `ot_dif_bindings.rs` looks like but I guess I need to wait until this PR lands :)

silvestrst replied:
It looks better with follow up changes from [[WIP] Improve generated DIF bindings](https://github.com/lowRISC/opentitan/pull/2154).

alphan commented on sw/device/tock/dif/build.rs:
```diff
@@ -14,10 +14,25 @@ fn main() {
+    let dif_libs = &["uart_ot"];
+    libraries_link(dif_lib_path.to_str().unwrap(), dif_libs);
+
+    let mmio_lib_path = Path::new(&build_root).join("sw/device/lib/base");
+    let base_libs = &["mmio_ot"];
+    libraries_link(mmio_lib_path.to_str().unwrap(), base_libs);
```
Really minor, but here too.

silvestrst replied:
Done.

silvestrst commented:
Hey @alphan and @mcy if there is nothing major wrong with these changes, it would be nice to merge this change - any further clean-ups we could do as follow-ups. I addressed all of the comments that are relevant to these changes, and left @mcy comments on the inherited code from upstream for the future cleanups.

alphan commented:
> Hey @alphan and @mcy if there is nothing major wrong with these changes, it would be nice to merge this change - any further clean-ups we could do as follow-ups. I addressed all of the comments that are relevant to these changes, and left @mcy comments on the inherited code from upstream for the future cleanups.

Thanks for updating this @silvestrst. LGTM.

silvestrst commented:
Thank you guys for the review!
